sdks/python/apache_beam/runners/direct/direct_runner.py

```diff
@@ -42,13 +44,91 @@
 from apache_beam.runners.runner import PipelineRunner
 from apache_beam.runners.runner import PipelineState
 from apache_beam.transforms.core import CombinePerKey
+from apache_beam.transforms.core import CombineValuesDoFn
 from apache_beam.transforms.core import ParDo
 from apache_beam.transforms.core import _GroupAlsoByWindow
 from apache_beam.transforms.core import _GroupAlsoByWindowDoFn
 from apache_beam.transforms.core import _GroupByKeyOnly
 from apache_beam.transforms.ptransform import PTransform
 
-__all__ = ['DirectRunner']
+# Note that the BundleBasedDirectRunner and SwitchingDirectRunner names are
+# experimental and have no backwards compatibility guarantees.
+__all__ = ['BundleBasedDirectRunner',
+           'DirectRunner',
+           'SwitchingDirectRunner']
+
+
+class SwitchingDirectRunner(PipelineRunner):
+  """Executes a single pipeline on the local machine.
+
+  This implementation switches between using the FnApiRunner (which has
+  high throughput for batch jobs) and using the BundleBasedDirectRunner,
+  which supports streaming execution and certain primitives not yet
+  implemented in the FnApiRunner.
+  """
+
+  def run_pipeline(self, pipeline):
+    use_fnapi_runner = True
+
+    # Streaming mode is not yet supported on the FnApiRunner.
+    if pipeline.options.view_as(StandardOptions).streaming:
+      use_fnapi_runner = False
+
+    from apache_beam.pipeline import PipelineVisitor
+    from apache_beam.runners.common import DoFnSignature
+    from apache_beam.runners.dataflow.native_io.iobase import NativeSource
+    from apache_beam.runners.dataflow.native_io.iobase import _NativeWrite
+    from apache_beam.testing.test_stream import TestStream
+
+    class _FnApiRunnerRunnabilityVisitor(PipelineVisitor):
+      """For internal use only; no backwards-compatibility guarantees.
+
+      Visitor for determining whether a Pipeline can be run on the FnApiRunner.
+      """
+
+      def __init__(self):
+        self.runnable = True
+
+      def visit_transform(self, applied_ptransform):
+        transform = applied_ptransform.transform
+        # The FnApiRunner does not support streaming execution.
+        if isinstance(transform, TestStream):
+          self.runnable = False
+        # The FnApiRunner does not support reads from NativeSources.
+        if (isinstance(transform, beam.io.Read) and
+            isinstance(transform.source, NativeSource)):
+          self.runnable = False
+        # The FnApiRunner does not support the use of _NativeWrites.
+        if isinstance(transform, _NativeWrite):
+          self.runnable = False
+        if isinstance(transform, beam.ParDo):
+          dofn = transform.dofn
+          # The FnApiRunner does not support execution of SplittableDoFns.
+          if DoFnSignature(dofn).is_splittable_dofn():
+            self.runnable = False
+          # The FnApiRunner does not support execution of CombineFns with side
```

Comment:
<!--thread_id:cc_167021935_t; commit:a2cf7f5de11964d3cb11a35665e8a8fd68b01d95; resolved:1-->
<!--section:context-quote-->
> **robertwb** wrote:
> with deferred side inputs

<!--section:body-->
Done.

Reply:
Not done?

Reply:
Oops, done.